Take constants into account

cpansprout opened the pull request:
Snooping in stashes to see whether there is a typeglob with a CODE
entry will fail for constants created by ‘use constant’, which are
stored shorthand as scalar references.

In addition, perl 5.28 will introduce an optimisation whereby most
subs are stored as code refs in stashes, without needing a typeglob.
This commit gets all tests passing with bleadperl.

Commit:
Take constants into account

Snooping in stashes to see whether there is a typeglob with a CODE
entry will fail for constants created by ‘use constant’, which are
stored shorthand as scalar references.

In addition, perl 5.28 will introduce an optimisation whereby most
subs are stored as code refs in stashes, without needing a typeglob.
This commit gets all tests passing with bleadperl.

belden reviewed (commented):
Could you add a test to show that the original `HasConstant::foo` gets rolled back properly once `$rs` leaves scope?

Commit:
Test constant restoration at scope end

Commit:
Suppress warnings on perl 5.14 and before

From perl5160delta:

       o   Redefinition warnings for constant subroutines used to be
           mandatory, even occurring under "no warnings".  Now they
           respect the warnings pragma.

Perl 5.14 and earlier still have the bug, so let’s work around it; but
take advantage of constant folding so that newer perls do not slow
down at all.

cpansprout commented:
Done.  But I also noticed it was warning with older perls, so I fixed that, too.

belden reviewed (approved):
Thanks for the updates - I haven't used been programming in Perl for almost two years now :/

belden commented:
I'll push updates to CPAN in a day or two.

cpansprout commented:
Thank you for the quick response.
